examples/single_page_app/main.py

```diff
@@ -1,34 +1,50 @@
 #!/usr/bin/env python3
-from router import Router
+from custom_sub_pages import custom_sub_pages, protected
 
-from nicegui import ui
+from nicegui import app, ui
 
 
-@ui.page('/')  # normal index page (e.g. the entry point of the app)
-@ui.page('/{_:path}')  # all other pages will be handled by the router but must be registered to also show the SPA index page
-def main():
-    router = Router()
+@ui.page('/')
+@ui.page('/{_:path}')
+def main_page():
+    with ui.header().classes('items-center bg-blue-100'):
+        ui.button('Home', on_click=lambda: ui.navigate.to('/')).props('flat')
+        ui.button('Secret', on_click=lambda: ui.navigate.to('/secret')).props('flat')
+        ui.button('Invalid', on_click=lambda: ui.navigate.to('/invalid')).props('flat')
+        ui.button('Error', on_click=lambda: ui.navigate.to('/error')).props('flat')
+        ui.space()
+        ui.button('Logout', icon='logout', on_click=lambda: (
+            app.storage.user.update(authenticated=False),
+            ui.navigate.to('/'),
+        )).props('flat').bind_visibility_from(app.storage.user, 'authenticated')
 
-    @router.add('/')
-    def show_one():
-        ui.label('Content One').classes('text-2xl')
+    custom_sub_pages({
+        '/': home,
+        '/secret': secret,
+        '/error': error,
+    }).classes('flex-grow p-4')
 
-    @router.add('/two')
-    def show_two():
-        ui.label('Content Two').classes('text-2xl')
 
-    @router.add('/three')
-    def show_three():
-        ui.label('Content Three').classes('text-2xl')
+def home():
+    ui.markdown('''
+        This example shows inheritance from `ui.sub_pages` for decorator-based route protection and a custom 404 page.
 
-    # adding some navigation buttons to switch between the different pages
-    with ui.row():
-        ui.button('One', on_click=lambda: router.open(show_one)).classes('w-32')
-        ui.button('Two', on_click=lambda: router.open(show_two)).classes('w-32')
-        ui.button('Three', on_click=lambda: router.open(show_three)).classes('w-32')
+        **Try it:** Navigate to "Secret" (passphrase: "spa") or "Invalid" for 404.
+    ''')
 
-    # this places the content which should be displayed
-    router.frame().classes('w-full p-4 bg-gray-100')
 
+def error():
+    raise ValueError('some error message')
 
-ui.run()
+
+@protected
```

Comment:
This decorator doesn't seem to work. My browser isn't asking for any credentials.